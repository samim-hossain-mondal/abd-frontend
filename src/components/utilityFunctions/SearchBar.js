import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { TextField, useMediaQuery } from '@mui/material';
import {useDebouncedCallback} from 'use-debounce';
import SearchIcon from '@mui/icons-material/Search';

export default function SearchBar({ query, setQuery }) {

  const [searchInput, setSearchInput] = useState(query.search);
  const aboveTablet = useMediaQuery('(min-width: 769px)');

  const debounced = useDebouncedCallback(
    (value) => {
      setSearchInput(value);
      setQuery({ ...query, search: value })
    },
    500,
  );

  const handleSubmit = (e) => {
    e.preventDefault();
    setQuery({ ...query, search: searchInput })
  }

  return (
    <form onSubmit={(e)=>{handleSubmit(e)}} onChange={(e)=>debounced(e.target.value)}>
      <TextField
        id="search-bar"
        className="text"
        label="Search"
        variant="outlined" placeholder="Search..." size="small"
        InputProps={{
          endAdornment: (
            <SearchIcon sx={{ color: 'primary.main' }}/>
          ),
        }}
        sx={{ width: aboveTablet ? '200px' : '138px' }}
      />
    </form>
  );
}
SearchBar.propTypes = {
  query: PropTypes.shape({
    search: PropTypes.string,
    status: PropTypes.string,
    startDate: PropTypes.string,
    endDate: PropTypes.string,
  }).isRequired,
  setQuery: PropTypes.func.isRequired
};