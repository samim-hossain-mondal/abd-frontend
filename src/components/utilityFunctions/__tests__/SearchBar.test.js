import React from 'react';
import {render,fireEvent,screen, waitFor} from '@testing-library/react';
import SearchBar from '../SearchBar';

describe('SearchBar', () => {
  it('should match snapshot', () => {
    const {container} = render(
      <SearchBar query={{search: '', status: '', startDate: '', endDate: ''}} setQuery={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should setQuery when submits', () => {
    const setQuery = jest.fn();
    render(
      <SearchBar query={{search: '', status: '', startDate: '', endDate: ''}} setQuery={setQuery} />
    );
    fireEvent.submit(screen.getByTestId('search-form'));
    expect(setQuery).toHaveBeenCalled();
  });

  it('should setQuery when changes', async () => {
    const setQuery = jest.fn();
    render(
      <SearchBar query={{search: '', status: '', startDate: '', endDate: ''}} setQuery={setQuery} />
    );
    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'test' } });
    await waitFor(() => {
      expect(setQuery).toHaveBeenCalled();
    });
  });
});