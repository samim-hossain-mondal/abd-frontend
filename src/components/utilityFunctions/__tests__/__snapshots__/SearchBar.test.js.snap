// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot 1`] = `
<div>
  <form
    data-testid="search-form"
  >
    <div
      class="MuiFormControl-root MuiTextField-root text css-1vuwe4s-MuiFormControl-root-MuiTextField-root"
      data-testid="search-bar"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="search-bar"
        id="search-bar-label"
      >
        Search
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
          id="search-bar"
          placeholder="Search..."
          type="text"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Search
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</div>
`;
